Changes

```diff
@@ -382,6 +382,16 @@ Working version
   in presence of a name collision.
   (Florian Angeletti, report by Thomas Refis, review by Gabriel Scherer)
 
+- #9974, #????: make `include` and with `constraints` handle correctly the
+  ghost components of signatures. For instance, in
+
+    include struct class c = object end end type c
+
+   the type `c` shadows the `class c` to avoid shadowing only the ghost type
+   c introduced by the class.
+  (Florian Angeletti and Gabriel Scherer, report by Eduardo Rafael,
+   review by ???)
```

Comment:
I would propose to move me away from the authors and into the reviewers. (We discussed the general direction beforehand but the code is all yours, except for the first trivial refactoring commit.)